Add Generated TMC TIM Topic

mwodahl opened the pull request:
# PR Details
## Description

This PR adds a new Kafka topic that contains TIMs generated by the TMC. This includes TIMs deposited via the TimDepositController, offloaded files, and those deposited through the TimReceiver.

## Motivation and Context

This update allows users to access a Kafka topic that only contains TMC-generated TIMs. 

## How Has This Been Tested?

This has been tested locally with docker-compose and Postman. TIMs were verified to be deposited to the new topic in the event they were generated by the TMC using kcat.

## Types of changes

- [ X ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ X ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ X ] I have read the **CONTRIBUTING** document.
[ODE Contributing Guide](https://github.com/usdot-jpo-ode/jpo-ode/blob/bugfix/Pull_request_template/docs/contributing_guide.md) 
- [ X ] I have added tests to cover my changes.
- [ X ] All new and existing tests passed.

payneBrandon commented on jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java:
```diff
@@ -202,6 +202,16 @@ public synchronized ResponseEntity<String> depositTim(String jsonString, Request
 
+      // If the TIM was generated by the TMC broadcast to the filtered TMC TIM Kafka topic
+      try {
+         String generatedBy = (new JSONObject(obfuscatedTimData)).getJSONObject("metadata").getString("recordGeneratedBy");
```
The TIMs coming into this endpoint will always be marked as TMC as this is the entry point for them to be generated. I'd like to discuss with the team on this one where it might make sense to pull the unique TMC messages from. If the system will be generating TIM messages solely through the ODE, it may make sense to only pull them from here and not in teh Asn1DecodeDataRouter. I think a good offline discussion will help us hash this one out a bit

mwodahl replied:
I've moved this message deposit from the Asn1DecodedDataRouter and the TimDepositController to the Asn1EncodedDataRouter after the phase 1 message encoding.

dmccoystephenson commented on jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/services/asn1/Asn1DecodedDataRouter.java:
```diff
@@ -103,6 +103,16 @@ public Object process(String consumedData) {
+					String generatedBy = (new JSONObject(odeTimData)).getJSONObject("metadata")
+							.getString("recordGeneratedBy");
+					if (generatedBy.equalsIgnoreCase("TMC_VIA_SAT") || generatedBy.equalsIgnoreCase("TMC_VIA_SNMP")) {
+						timProducer.send(odeProperties.getKafkaTopicOdeTimJsonTMCFiltered(), null, odeTimData);
```
Is it possible that we'll get duplicates in this new topic?

mwodahl replied:
This has been updated to send TMC-generated messages from one place instead of two so there won't be duplicate messages in the new topic

dmccoystephenson reviewed (approved):
Looks good & the unit tests pass!

payneBrandon reviewed (changes requested):
Discussed offline, but we'll want to tweak this to push the same TIM data as the topic.OdeTimJson, but with the additional asn1 field filled out. Since this is the encoded path, getting the combo is more challenging. One option we talked about is adding a key to the metadata (which is maintained through encoding) and using KTable on top of the ode tim topic to find the corresponding json and combine.

mwodahl commented on docs/data-flow-diagrams/tim/TIM Data Flow.drawio:
```diff
@@ -1,197 +1,197 @@
-<mxfile host="Electron" modified="2024-06-25T15:54:15.452Z" agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) draw.io/24.2.5 Chrome/120.0.6099.109 Electron/28.1.0 Safari/537.36" etag="pg-zYBY0A62ucJJm-_Ut" version="24.2.5" type="device">
```
Updated the diagram to reflect this

mwodahl commented:
Updated this to include an additional topic, KeyedOdeTimJson, that has a unique identifier attached to the TIM JSON that allows for later querying in the Asn1EncodedDataRouter. I added a new topic instead of reusing the OdeTimJson topic as it prevents unnecessary information from being included in the OdeTimJson topic and allows for pushing a null record with the generated TIM UUID to erase the TIM record from TimJson-Store after it's been retrieved. I've also updated the diagrams to reflect these changes.

payneBrandon commented on jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/services/asn1/Asn1EncodedDataRouter.java:
```diff
@@ -189,6 +201,9 @@ public void processEncodedTim(ServiceRequest request, JSONObject consumedObj) {
          // Sign and send to RSUs
 
+         // Deposit to filtered topic if TIM is TMC-generated
+         depositToFilteredTopic(metadataObj, dataObj, request);
```
Should we let it be signed if possible then deposit?

payneBrandon commented on jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/ServiceRequest.java:
```diff
@@ -101,6 +101,15 @@ public boolean equals(Object obj) {
    private SituationDataWarehouse.SDW sdw;
    private RSU[] rsus;
    private SNMP snmp;
+   private String uuid;
```
Any downside to having this as a UUID instead of String?

mwodahl replied:
I originally added that as a String so that the SerDes class could support it for the k-table without being overridden. This uuid field is no longer present in the ServiceRequest class as there is a unique streamId already present on the TIM that I'm using to identify it instead.

payneBrandon commented on jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/traveler/TimDepositController.java:
```diff
@@ -202,6 +203,18 @@ public synchronized ResponseEntity<String> depositTim(String jsonString, Request
       stringMsgProducer.send(odeProperties.getKafkaTopicOdeTimBroadcastJson(), null, obfuscatedTimData);
 
+      try {
+         String generatedBy = (new JSONObject(obfuscatedTimData)).getJSONObject("metadata").getString("recordGeneratedBy");
```
Should we be using the object sent down to the topic.OdeTimJson below? Not sure what the difference between that object and this broadcast object is, but we _have_ been pulling from the topic.OdeTimJson so keeping with that would probably be best

mwodahl replied:
This now uses the object in the OdeTimJson topic directly

dmccoystephenson commented on jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/OdeTimJsonTopology.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.Properties;
+
+
+public class OdeTimJsonTopology {
```
What is the purpose of this new class? A descriptive comment may be good here.

mwodahl replied:
Added a comment to describe the purpose of this class

dmccoystephenson reviewed (approved):
Looks good & the unit tests pass, I just had one question/suggestion

payneBrandon commented:
@mwodahl looks like we have a merge conflict from Matt's properties changes

Michael7371 reviewed (changes requested):
From my local testing, it looks like the `topic.OdeTimJson` still is missing the `asn1` field. Is this the expected behavior?

mwodahl commented:
> From my local testing, it looks like the `topic.OdeTimJson` still is missing the `asn1` field. Is this the expected behavior?

Yep, that's expected. The asn1 field will only be present in the topic.OdeTimJsonTMCFiltered topic.

dmccoystephenson commented on jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/services/asn1/Asn1EncodedDataRouter.java:
```diff
@@ -404,4 +419,32 @@ private String stripHeader(String encodedUnsignedTim) {
+      } catch (Exception e) {
+         logger.error("Error while fetching recordGeneratedBy field: {}", e.getMessage());
+      }
+      return null;
```
Is returning 'null' here necessary since the method has a return type of void? Also, should the return type be capitalized like it is currently?

mwodahl replied:
Swapped from "Void" to "void" and removed return - good catch!

dmccoystephenson reviewed (changes requested):
Just the one comment

Michael7371 reviewed (approved):
LGTM!
